src/v/archival/ntp_archiver_service.cc

```diff
@@ -1383,20 +1384,51 @@ ntp_archiver::schedule_uploads(std::vector<upload_context> loop_contexts) {
             _probe->upload_lag(ctx.last_offset - ctx.start_offset);
         }
 
-        while (uploads_remaining > 0 && may_begin_uploads()) {
-            auto should_stop = co_await ctx.schedule_single_upload(*this);
-            if (should_stop == ss::stop_iteration::yes) {
-                break;
-            }
+        std::exception_ptr ep;
+        try {
+            while (uploads_remaining > 0 && may_begin_uploads()) {
+                auto should_stop = co_await ctx.schedule_single_upload(*this);
+                if (should_stop == ss::stop_iteration::yes) {
+                    break;
+                }
 
-            // Decrement remaining upload count if the last call actually
-            // scheduled an upload.
-            if (!ctx.uploads.empty()) {
-                const auto& last_scheduled = ctx.uploads.back();
-                if (last_scheduled.result.has_value()) {
-                    uploads_remaining -= 1;
+                // Decrement remaining upload count if the last call actually
+                // scheduled an upload.
+                if (!ctx.uploads.empty()) {
+                    const auto& last_scheduled = ctx.uploads.back();
+                    if (last_scheduled.result.has_value()) {
+                        uploads_remaining -= 1;
+                    }
+                }
+            }
+        } catch (...) {
+            ep = std::current_exception();
+        }
+        if (ep) {
+            _rtclog.warn("Failed to schedule upload: {}", ep);
+            std::vector<ss::future<>> pending_uploads;
+            pending_uploads.reserve(
+              ctx.uploads.size() + scheduled_uploads.size());
+            for (auto& upload : ctx.uploads) {
+                if (upload.result.has_value()) {
+                    pending_uploads.emplace_back(
+                      std::move(upload.result.value()).discard_result());
+                }
+            }
+            for (auto& scheduled : scheduled_uploads) {
```

Comment:
i'm still not seeing how the input `loop_contexts` is cleaned up (or does it not need to be?)

Reply:
oh, i see this is changed in the next commit. can you sequence the simplification commit before the commit with the fix?

Reply:
Done.